src/CompletionProvider.php

```diff
@@ -169,122 +197,112 @@ public function provideCompletion(PhpDocument $doc, Position $pos): CompletionLi
-                    } else if ($namespace && !($prefix && $isFullyQualified)) {
-                        // Insert the global FQN with trailing backslash
-                        $item->insertText = '\\' . $fqn;
-                    } else {
-                        // Insert the FQN without trailing backlash
-                        $item->insertText = $fqn;
+                $fqnStartsWithPrefix = strStartsWith($fqn, $prefix);
+                $fqnContainsPrefix = empty($prefix) || strpos($fqn, $prefix) !== false;
+                if (($def->canBeInstantiated || ($def->isGlobal && !isset($creation))) && $fqnContainsPrefix) {
+                    if ($namespaceDefinition !== null && $namespaceDefinition->name !== null) {
+                        $namespacePrefix = (string)PhpParser\ResolvedName::buildName($namespaceDefinition->name->nameParts, $node->getFileContents());
+
+                        $isAliased = false;
+
+                        $isNotFullyQualified = !($class instanceof Node\QualifiedName) || !$class->isFullyQualifiedName();
+                        if ($isNotFullyQualified) {
+                            foreach ($namespaceImportTable as $alias => $name) {
+                                if (strStartsWith($fqn, $name)) {
+                                    $fqn = $alias;
+                                    $isAliased = true;
+                                    break;
+                                }
+                            }
+                        }
+
+
+                        if (!$isNotFullyQualified && ($fqnStartsWithPrefix || strStartsWith($fqn, $namespacePrefix . "\\" . $prefix))) {
+                            // $fqn = $fqn;
+                        } elseif (!$isAliased && !array_search($fqn, array_values($namespaceImportTable))) {
+                            if (empty($prefix)) {
+                                $fqn = '\\' . $fqn;
+                            } elseif (strStartsWith($fqn, $namespacePrefix . "\\" . $prefix)) {
+                                $fqn = substr($fqn, strlen($namespacePrefix) + 1);
+                            } else {
+                                continue;
+                            }
+                        } elseif (!$isAliased) {
+                            continue;
+                        }
+                    } elseif ($fqnStartsWithPrefix && $class instanceof Node\QualifiedName && $class->isFullyQualifiedName()) {
+                        $fqn = '\\' . $fqn;
                     }
+
+                    $item = CompletionItem::fromDefinition($def);
+
+                    $item->insertText = $fqn;
                     $list->items[] = $item;
                 }
             }
-            // Suggest keywords
-            if ($node instanceof Node\Name && $node->getAttribute('parentNode') instanceof Node\Expr\ConstFetch) {
+
+            if (!isset($creation)) {
                 foreach (self::KEYWORDS as $keyword) {
```

Comment:
While yes, that could be done, there are already enough changes in this PR and that would have nothing to do with the parser adoption.